### DISCUSSION TITLE
[v3] Form object hydrate rules into component before mount

### DISCUSSION THREAD

**Discussion by DanielSpravtsev** [livewire/livewire#6174] [7. Bugs (archived)]
### Livewire version

v3.0.0-beta.6

### Which PHP version are you using?

PHP 8.2

### Steps To Reproduce

Component form rules injects before component mount sets id property, which value should be used inside UserForm rules function.
After call submit(), it returns 'unique value exists error', because id property is null when it was mounted.

@viliusvsx Does your PR https://github.com/livewire/livewire/pull/5904 covers this issue? 

```
<?php

namespace App\Livewire;

use App\Livewire\Forms\PostForm;
use Livewire\Component;

class Post extends Component
{
    public PostForm $form;

    public function submit()
    {
        $this->form->save();
    }

    public function mount()
    {
        $this->form->fillFormWithDummyData();
    }

    public function render()
    {
        return view('livewire.post');
    }
}
```

```
<?php

namespace App\Livewire\Forms;

use Livewire\Attributes\Rule;
use Livewire\Form;

class PostForm extends Form
{
    public ?string $id = null;

    public function rules()
    {
        return [
            'imaginable_field' => 'rule_like_unique:table_name,column_name,'.$this->id
        ];
    }
    public function save()
    {
        /*
        When we call
        $this->validate()
        It will validate components rules when postForm wasn't filled with component mount
        But we can call postForm rules to ensure that id value is presented
        */

        dd([
            'component rules' => $this->getComponent()->getRules(),
            'postForm rules' => $this->rules(),

        ]);
    }

    public function fillFormWithDummyData(): void
    {
        $this->id = uuid_create();
    }
}
```

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [x] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by DanielSpravtsev** [comment]
Also this discussion references to https://github.com/livewire/livewire/discussions/6058 for more examples

**Comment by viliusvsx** [comment]
I think, that id always be null, because you never set it with value

**Reply by DanielSpravtsev** [reply]
@viliusvsx This modal used for creating/updating user form, so in component ```mount()``` method it is filling with ```fill``` method when it passed from frontend. 

$this->form->fill(User::whereId($data['modelId'])->first()->toArray()) simple set model array values to form properties when they exists (model id => form id and etc)

For 100% confidence I can call in form create method dd of $this->getComponent()->getRules() and $this->id and see that id is filled and not presented in rules because of component rules hydrates **before** id filled in mount and for that reason id in rules is null

Hope you get it :)

**Reply by m15studio** [reply]
@viliusvsx I confirm that the rules hydrates before id property filled by component object

**Reply by viliusvsx** [reply]
@DanielSpravtsev Is it possible to get your repo? I want to clone this and test for specific your case

**Reply by DanielSpravtsev** [reply]
@viliusvsx https://github.com/DanielSpravtsev/livewire3-form-objects

**Comment by joshhanley** [comment]
@DanielSpravtsev hmm I would expect that rule to fail as there is no provided ID.

I would just add it as a conditional rule:

```php
public function rules()
{
    $rules = [
        'id',
    ];
    
    if ($this->id !== null) {
        $rules = $rules + [
            'some_field' => 'unique:table,column,'.$this->id
        ];
    }
    
    return $rules;
}
```

**Reply by DanielSpravtsev** [reply]
It's not about it. If you use rule with conditional property (that can be presented / not presented) in Livewire component, everything will work fine, but when I use same rule in Form rules, it will not work, because Form Rules don't merge into Livewire rules after property updated, only at first livewire component hydration, except when using rules at Livewire component itself. Is it not a bug? It is strange behaviour

**Reply by DanielSpravtsev** [reply]
If you look at my repository with this issue, you could see that I filling form object values at ```mount()```, but livewire component merges Form object rules before I filled it with values. 

As I think, it should merge rules to Livewire after mount (if I using form object), so I don't need to merge it manually after filling values at mount

**Reply by raustin-m** [reply]
I think this is also causing a similar issue for me. I'm trying to validate one field based on another (actually on a property that isn't a form field but just a regular public property that's set on the form when the field is changed), and this works in the component `rules()` but does not in the form's `rules()`:

```php
    // in the component
    protected function rules()
    {
        return [
            'form.postcode' => ['required', new PostalCode($this->form->country)],
        ];
    }
```

When this is in the form's rules (minus the prefixed key and extra `->form`), the `country` property is never populated, but in the component's `rules()` it's always populated with the latest value from the form.  I suspected this is for the same reason @DanielSpravtsev just mentioned, that the rules are being added to the component from the form before the form is hydrated.

The only way I can get it to work for now is if I pass a closure to my validation rule:
```php
    // in the form
    protected function rules()
    {
        return [
            'postcode' => ['required', new PostalCode(fn () => $this->country)],
        ];
    }
```

Then the rule can get the actual value, but it's not an ideal refactor and I kinda wish it worked like it did in the component.
